optimade/server/entry_collections/ase.py

```diff
@@ -0,0 +1,191 @@
+    ) -> Generator[RowInfo, None, None]:
+        """Select rows using a filter function."""
+        for i, row in enumerate(self.rows):
+            if i == skip + limit:
+                return
+            if i >= skip and filter_function(row):
+                yield row
+
+    def count(self, filter_function, skip=0, limit=INF):
+        return sum(1 for _ in self.select(filter_function, skip=skip, limit=limit))
+
+    def get_result(self, id: int) -> Dict[str, Any]:
+        row = self.db.get(id=id)
+        count = dict(sorted(row.count_atoms().items()))
+        formula = Formula.from_dict(count)
+        abc, _, N = formula.stoichiometry()
+        anonymous = dict2str({symb: n * N for symb, n in abc._count.items()})
+        t = row.mtime * YEAR + T2000
+        last_modified = strftime("%Y-%m-%dT%H:%M:%SZ", gmtime(t))
+        dct = {
+            "cartesian_site_positions": row.positions.tolist(),
+            "species_at_sites": row.symbols,
+            "nsites": row.natoms,
+            "species": [
+                {"name": symbol, "chemical_symbols": [symbol], "concentration": [1.0]}
+                for symbol in count
+            ],
+            "lattice_vectors": row.cell.tolist(),
+            "dimension_types": row.pbc.astype(int).tolist(),
+            "last_modified": last_modified,
+            "elements": list(count),
+            "nelements": len(count),
+            "elements_ratios": [n / row.natoms for n in count.values()],
+            "chemical_formula_descriptive": "x",
+            "chemical_formula_reduced": f"{formula}",
+            "chemical_formula_hill": f"{formula:hill}",
+            "chemical_formula_anonymous": anonymous,
+            "nperiodic_dimensions": sum(row.pbc),
+            "structure_features": [],
+            "id": str(row.id),
+        }
+
+        return dct
+
+
+def create_code_string(query: ASEQueryTree) -> str:
+    """Convert query-tree object to a Python function.
+
+    Example:
+
+    >>> tree = ('AND', [('nelements', ('=', 3)),
+    ...                 ('elements', ('HAS', '"Cu"'))])
+    >>> create_code_string(tree)
+    <function <lambda> at 0x7f6d5bad6ca0>
+    """
+
+    q1, q2 = query
+
+    if q1 == "AND":
+        return " and ".join(f"({create_code_string(q)})" for q in q2)
+
+    if q1 == "nelements":
+        op, val = q2
+        if op == "=":
+            op = "=="
+        return f"len(row.elements) {op} {val}"
+
+    if q1 == "elements":
+        q3, q4 = q2
+
+        if q3 == "HAS":
+            return f"{q4} in row.elements"
+
+        if q3 == "HAS ALL":
+            return f"all(symbol in row.elements for symbol in {q4})"
+
+    raise NotImplementedError(query)
+
+
+@lru_cache(maxsize=100)
+def create_filter_function(query: ASEQueryTree) -> FilterFunction:
+    """Convert query-tree from ASETransofmer to a Python filter-function."""
+    code_string = create_code_string(query)
+    code = compile("lambda row: " + code_string, "<string>", "eval")
```

Comment:
You have to be very careful when using eval in combination with a user supplied string. By crafting an ingenious query, a user can potentially execute code on your machine.
What would, for example, happen when I give the query ('elements', ('HAS', '"Cu", print("Executing user inserted code")')). Will the print statement be executed?

Reply:
Good point.  I'll keep that in mind.  The query string from the user goes through the Lark-parser and the OPTIMADE grammar, so hopefully that narrows down the possibilities.

Reply:
This is indeed why I want to push this into the filter transformer instead. I have time to take a stab at it from Wednesday and forward.

Reply:
Hmm, sorry for the delay here - this month has proven to be a bit more troublesome than anticipated (moving to Norway at the end of the month), but I have already started working on this a bit locally, and I'm starting to see where you were hitting walls. I still think a post-processing in the filter transformer class (to begin with) will be a better option for all of this though. So I'll keep tinkering locally when I have time to spare.

Reply:
No problem.  I'll wait for your changes and then go on with implementing more of the filter language syntax.